### PR TITLE
GRPO PrefixGrouper: shared-prefix grad path in grpo_accumulated_loss

### PR DESCRIPTION
**Pull request by danielhanchen**
### What

Companion to the PrefixGrouper change in unslothai/unsloth. This PR adds the grad-path half in `grpo_accumulated_loss`.

In GRPO every prompt spawns G = num_generations completions that share the prompt prefix. The full-row packed path forwards that prefix G times; PrefixGrouper stores it once and concatenates only the G suffixes (a shared-prefix attention mask), cutting the training forward from `G*(P+R)` to `P+G*R` tokens. It is mathematically identical: the shared prefix contributes the same gradient as the sum of the G repeats.

### This PR

- Adds a PrefixGrouper branch to `grpo_accumulated_loss` producing grad-carrying `new_logprobs`, gated behind `UNSLOTH_GRPO_PREFIX_GROUPER` (default on; set 0 to disable). Auto-disabled under vLLM (`fast_inference=True`).
- First-use self-verify against the full-row packed path (fall back and mark the shape unsafe on mismatch), so a mask or isolation regression cannot ship silently.
- The kernel and layout builder live in unsloth (`unsloth.utils.prefix_grouper`), imported lazily and guarded, so this change is a safe no-op if it merges before the unsloth companion.

Requires the unsloth companion PR for the kernel and layout builder; a safe no-op if it merges before the companion (the lazy import degrades to off).

### Result

2.55x faster and 3.37x lower-memory training forward at favorable prompt:completion ratios. Correctness is fp64 bit-exact; parity, backward, and GQA verified. Numbers in a comment below.

### CODE REVIEW

**Comment by danielhanchen** [comment]
### Benchmark and correctness

Notation: for a group of G completions sharing a prompt of length P with mean completion length R, let `tok_r = G*(P+R) / (P + G*R)`. PrefixGrouper forwards `P + G*R` tokens instead of `G*(P+R)`, so memory tracks `tok_r` and the trunk forward speeds up with it.

Correctness (Qwen2.5-0.5B and a real-model check):

- FlexAttention shared-prefix kernel vs the SDPA oracle and the per-row forward: fp64 bit-exact, fp32 <= 6.5e-5, bf16 first-completion-token bit-exact.
- Backward vs the oracle: fp64 ~1e-15, fp32 ~1e-6. GQA (`enable_gqa`) equals manual KV-repeat, bit-exact.
- Negative controls (a leaky cross-suffix mask, a truncated prefix) are both detected at 4-5 nats, so the parity test has teeth.

Isolated training forward + backward on a real model (Qwen3-4B, `tok_r = 3.33`):

| metric | full-row | PrefixGrouper | ratio |
| --- | --- | --- | --- |
| fwd+bwd time | 1.00x | 0.39x | 2.55x faster |
| peak memory | 1.00x | 0.30x | 3.37x less |

Kernel micro-benchmark (attention op, swept over P, R, G): memory always wins and tracks `tok_r`; the speed win grows with `tok_r`. Long shared prompt with big G (P=1024, R=256, G=16): 1.93x forward, 1.99x fwd+bwd, 3.98x memory.

Honest note on end-to-end: with a colocated vLLM rollout (`fast_inference`), generation is 77-94% of a GRPO step and PrefixGrouper does not touch it, so the end-to-end step speedup on a short-prompt notebook is small (about 1.03-1.04x) and it is memory-first there. The win is real end-to-end for long-prompt / large-G / long-context GRPO, for `num_iterations > 1`, and on the multi-GPU / FSDP2 path where the training forward is on the critical path. Default off, and auto-gated by `tok_r` so it never slows a low-`tok_r` batch.

**Comment by danielhanchen** [comment]
Companion unsloth PR (kernel, layout, attention routing, no-grad path): unslothai/unsloth#6871. This grad-path change is a safe no-op without it (the unsloth import is guarded and falls back to the full-row path).

File: unsloth_zoo/rl_replacements.py [hunk 1]
```diff
@@ -944,6 +944,77 @@ def slice_sample_axis(value, start, end):
...
+    # ---- PrefixGrouper (GRPO shared-prompt dedup; default OFF => byte-identical) ----
+    # In GRPO every prompt spawns G=num_generations completions that share the prompt prefix.
+    # The full-row packed path below forwards that prefix G times; PrefixGrouper stores it ONCE
+    # and concatenates only the G suffixes (FlexAttention shared-prefix mask), cutting the trunk
+    # forward from G*(P+R) to P+G*R tokens. Gated behind UNSLOTH_GRPO_PREFIX_GROUPER (requires
+    # seq-packing on). tok_r auto-gate + first-use self-verify vs the full-row packed new_logprobs
+    # (fall back + mark-unsafe on mismatch). Loss/gradients flow through the shared-prefix stream
+    # (the prefix contributes grad once = the sum of the G repeats, mathematically identical). When
+    # off / grouping fails / not yet verified, the full-row packed path below runs as before.
+    _pg_result = None
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py:985]: Good catch, this was real: the packing block's local import made the name function-local, so the PrefixGrouper fallback handler could hit UnboundLocalError before the import ran. Fixed in 282b249 by moving the single local import to the top of the function (it stays local because this function's source is copied into the generated trainer without module imports) and reverting the unused module-level import.

File: unsloth_zoo/rl_replacements.py [hunk 2]
```diff
@@ -26,7 +26,7 @@
 import numpy as np
 from typing import Union, Callable, Optional, List, Dict
 from .device_type import DEVICE_TYPE, device_synchronize
-from .temporary_patches.common import torch_compile_options
+from .temporary_patches.common import torch_compile_options, UNSLOTH_ENABLE_LOGGING
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: Removed in 282b249 (the name is imported locally inside grpo_accumulated_loss, which shadowed this module-level import).

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py [hunk 3]
```diff
@@ -1078,7 +1186,79 @@ def slice_sample_axis(value, start, end):
...
+                input_ids = _pg_layout.flat_ids,
+                position_ids = _pg_layout.position_ids,
+                prefix_seg_info = _pg_layout.prefix_seg_info,
+                use_cache = False,
+            ).logits
+            return _pg_layout.extract_logps(
+                _h, lm_head, chunked_hidden_states_selective_log_softmax,
+                _pg_chunks, logit_scale_multiply, logit_scale_divide,
+                logit_softcapping, temperature,
+            )  # [total_rows, W] with grad
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: Added the same device_synchronize guard inside the PG forward.

File: unsloth_zoo/rl_replacements.py [hunk 4]
```diff
@@ -1078,7 +1186,79 @@ def slice_sample_axis(value, start, end):
...
+            _pg_use = False
+            os.environ["UNSLOTH_RETURN_HIDDEN_STATES"] = "1"
+            if isinstance(_pg_err2, torch.cuda.OutOfMemoryError):
+                torch.cuda.empty_cache()
+            if UNSLOTH_ENABLE_LOGGING:
+                print(f"[Unsloth] GRPO PrefixGrouper (grad) forward failed -> full-row packed: {_pg_err2!r}", flush = True)
+
+    if _pg_use and _pg_result is not None:
+        new_logprobs = _pg_result            # PrefixGrouper verified/trusted -> skip the loop
+        zipped_inputs = []
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py:1295]: Now freed right after the PG grad forward succeeds. Kept alive until then so a PG failure can still fall back to the packed result instead of the padded loop.

File: unsloth_zoo/rl_replacements.py [hunk 5]
```diff
@@ -944,6 +944,117 @@ def slice_sample_axis(value, start, end):
...
+            if _pg_layout is not None and _pg_layout.signature in _pg_unsafe:
+                _pg_layout = None
+            elif _pg_layout is not None:
+                _pg_layout.W = logits_to_keep + max_left_pad
+                _pg_verified = getattr(unwrapped_model, "_unsloth_prefix_grouper_grad_verified", None)
+                if _pg_verified is None:
+                    _pg_verified = set()
+                if (not _pg_verify_on()) or _pg_layout.signature in _pg_verified:
+                    _pg_trusted = True
+                    _pg_skip_pack = True   # trusted shape -> skip the full-row forward
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: By design: running the full-row packed forward on every trusted step would erase the savings PG exists for. The padded loop is a correct fallback for that rare step; fixed the log message that wrongly promised a packed fallback there.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py [hunk 6]
```diff
@@ -1078,7 +1186,85 @@ def slice_sample_axis(value, start, end):
...
+                # else: no full-row reference (packing off/failed) -> cannot verify -> fall back.
+            if _pg_trusted:
+                _pg_result = _pg_grad_forward()   # grad forward for the loss
+                _pg_use = True
+                # PG owns the loss now: drop the full-row packed graph from the verify step
+                # (kept until here so a PG forward failure could still fall back to it)
+                _pack_hidden = _pack_sel = _pack_result = None
+        except Exception as _pg_err2:
+            _pg_use = False
+            os.environ["UNSLOTH_RETURN_HIDDEN_STATES"] = "1"
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py:1275]: The failure handler now untrusts the signature so the next batch runs the packed path again, and marks it unsafe on OOM since that would recur at those lengths.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py [hunk 7]
```diff
@@ -944,6 +944,134 @@ def slice_sample_axis(value, start, end):
...
+    # the cache (the function object is rebuilt there once per module and grpo_accumulated_loss
+    # is a cache-module global), so memoize on grpo_accumulated_loss itself. The env gate is
+    # checked FIRST so the default-off path never imports or runs any PG code (byte-identical
+    # when unset). () means gate off / unavailable / failed import -> PrefixGrouper stays off.
+    _pg_funcs = getattr(grpo_accumulated_loss, "_pg_funcs", None)
+    if _pg_funcs is None:
+        _pg_funcs = ()
+        if os.environ.get("UNSLOTH_GRPO_PREFIX_GROUPER", "1").lower() not in (
+            "0", "false", "no", "off",
+        ):
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py:960]: Default-on is intentional here. PrefixGrouper stays auto-disabled under vLLM (fast_inference=True) and by the softcap/SSM/tok_r gates, and the first-use self-verify falls back on any mismatch, so it is a memory-first default on the raw-transformers path with no correctness change. Set UNSLOTH_GRPO_PREFIX_GROUPER=0 to opt out.

File: unsloth_zoo/rl_replacements.py [hunk 8]
```diff
@@ -1078,7 +1203,101 @@ def slice_sample_axis(value, start, end):
...
+                        _pg_u = getattr(unwrapped_model, "_unsloth_prefix_grouper_grad_unsafe", None)
+                        if _pg_u is None:
+                            _pg_u = set()
+                        if _pg_diff >= _PG_TOL_KILL:
+                            _pg_u.add(_pg_layout.signature)
+                            unwrapped_model._unsloth_prefix_grouper_grad_unsafe = _pg_u
+                        _pg_trusted = False
+                # else: no full-row reference (packing off/failed) -> cannot verify -> fall back.
+            if _pg_trusted:
+                _pg_result = _pg_grad_forward()   # grad forward for the loss
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: Good catch. Now freeing the packed hidden/sel/result before the PG grad forward (it is already consumed as the verify reference above), so PG can reuse that memory instead of holding both graphs; on a PG failure the padded loop recomputes new_logprobs.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py [hunk 9]
```diff
@@ -1078,7 +1203,103 @@ def slice_sample_axis(value, start, end):
...
+
+    if _pg_layout is not None:
+        try:
+            if not _pg_trusted:
+                # first use for this structure: verify vs the full-row packed new_logprobs
+                # (itself self-verified vs per-row). PASS < tol_ok -> trust; >= TOL_KILL ->
+                # mark unsafe forever; borderline -> fall back this shape.
+                if _pack_use and _pack_result is not None:
+                    with torch.no_grad():
+                        _pg_ref = _pg_grad_forward()
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py:1231]: Good catch. The verify forward runs while the full-row packed graph is still resident (it is the comparison reference), so an OOM there does not prove PG alone cannot fit. Fixed in 04ac1bf: a _pg_phase_verify flag now gates the mark-unsafe, so only an OOM in the PG-alone forward (after the packed graph is freed) marks the signature unsafe. A verify-phase OOM just falls back and retries the full-row packed path next batch. The cache is still emptied on every OOM.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py [hunk 10]
```diff
@@ -1078,7 +1186,85 @@ def slice_sample_axis(value, start, end):
...
+                        print(
+                            f"[Unsloth] GRPO PrefixGrouper (grad) verify: sig={_pg_layout.signature} "
+                            f"shared-prefix vs full-row-packed max|d|={_pg_diff:.4f}", flush = True,
+                        )
+                    if _pg_diff < _pg_tol_ok():
+                        _pg_v = getattr(unwrapped_model, "_unsloth_prefix_grouper_grad_verified", None)
+                        if _pg_v is None:
+                            _pg_v = set()
+                        _pg_v.add(_pg_layout.signature)
+                        unwrapped_model._unsloth_prefix_grouper_grad_verified = _pg_v
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: Correct, and handled: the grad-forward except handler pops the signature from the verified dict on any failure, and only adds it to the unsafe set on a PG-alone OOM (after the packed graph is freed, gated by not _pg_phase_verify). A failed grad forward never leaves the shape trusted, so the next batch reruns the packed path and re-verifies.

File: unsloth_zoo/rl_replacements.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py]: Handled by the verified envelope: trust requires _pg_T <= env[0] and _pg_maxseg <= env[1], and the envelope only widens on a passing verify. A longer T or segment fails the check, keeps packing on, and re-verifies.

**Comment by danielhanchen** [comment]
## Benchmarks

All numbers are Qwen3-4B LoRA GRPO on a single H200-class GPU, `num_generations = 8`, completion length 256, measured with a steady-state `StatisticsCallback` (median over 11 steady steps), so first-step compile/warmup is excluded. PrefixGrouper (PG) only touches the policy/reference logprob forward, so the relevant axis is the update-step time and the training-forward peak memory, not generation.

### 1. FlexAttention recompilation (the enabler for default-on)

GRPO changes the packed length `T` almost every batch. Compiling the FlexAttention kernel with `dynamic = False` keys the compiled artifact on shape, so a new `T` recompiled from scratch every step. Switching to `dynamic = True` compiles twice during warmup and is then recompile-free on new shapes.

| kernel compile | first call on a NEW T_pad | repeat call (same shape) | new/repeat ratio |
| --- | --- | --- | --- |
| `dynamic = False` | 14672 ms | 0.4 ms | 41681x |
| `dynamic = True` | 1.4 ms | 0.4 ms | 3.5x |

After the two warmup compiles, every subsequent new shape is ~1.3-1.5 ms with `dynamo_compiles += 0`. A different mask layout at the same `T_pad` does not recompile (0.3-0.7 ms), so only the padded length matters.

### 2. End-to-end GRPO step (PG on vs off)

| config | update step | median step | peak memory |
| --- | --- | --- | --- |
| `fast_inference=False`, PG off | 415.7 ms | 452.1 ms | 10.19 GB |
| `fast_inference=False`, PG on | 314.5 ms | 333.3 ms | 8.54 GB |
| `fast_inference=True`, PG off | 374.0 ms | 378.8 ms | 116.8 GB |
| `fast_inference=True`, PG on | 361.6 ms | 2084.5 ms | 121.5 GB |

- On the transformers path (`fast_inference=False`) PG cuts the logprob-forward update step by 1.32x (415.7 to 314.5 ms) and peak training-forward memory by ~16% (10.19 to 8.54 GB).
- Under a colocated vLLM rollout (`fast_inference=True`) the memory is dominated by vLLM's reservation and the first-use self-verify runs the full-row path too, so PG is net negative there. It is therefore auto-disabled when `use_vllm` is set (the numbers above confirm the gate is the right call).

### 3. Correctness / parity

- First-use self-verify of the shared-prefix forward vs the full-row packed reference over the exact loss mask: `max|d| = 0.24` nats (grad path), well under the `0.7` trust tolerance; re-verified whenever `T` or the longest segment grows.
- Isolated kernel vs the SDPA oracle and a per-row forward: fp64 bit-exact, fp32 <= 6.5e-5, bf16 first-completion-token bit-exact. Backward fp64 ~1e-15, fp32 ~1e-6; GQA equals manual KV-repeat bit-exact. Negative controls (a leaky cross-suffix mask, a truncated prefix) are detected at 4-5 nats.

### 4. Isolated training forward + backward (where the trunk forward is on the critical path)

For a group of `G` completions sharing a prompt of length `P` with mean completion length `R`, let `tok_r = G*(P+R) / (P + G*R)`; PG forwards `P + G*R` tokens instead of `G*(P+R)`.

| Qwen3-4B, `tok_r = 3.33` | full-row | PrefixGrouper | ratio |
| --- | --- | --- | --- |
| fwd+bwd time | 1.00x | 0.39x | 2.55x faster |
| peak memory | 1.00x | 0.30x | 3.37x less |

Kernel micro-benchmark, long shared prompt with big G (P=1024, R=256, G=16): 1.93x forward, 1.99x fwd+bwd, 3.98x memory. Memory always wins and tracks `tok_r`; the speed win grows with `tok_r`.

The win is real end-to-end for long-prompt / large-G / long-context GRPO, for `num_iterations > 1`, and on the multi-GPU / FSDP path where the training forward is on the critical path. On short-prompt vLLM runs it is memory-first, which is why it is gated off under vLLM. Auto-gated by `tok_r` so it never slows a low-`tok_r` batch.

**Comment by danielhanchen** [comment]
@codex review
